_drafts/debug-vs-release-builds.md

```diff
@@ -0,0 +1,239 @@
+release builds: we end up with a **different** executable:
+
+- Sizes will likely be different: stack or heap usage (affecting program
+  stability!), code space (affecting flash memory allocation, OTA)
+- Different machine code executed by the CPU: this can impact **timing** (often
+  where the nastiest bugs show up) and actual **program validity**, e.g.
+  assumptions you are making about certain spots in your program may no longer
+  be valid with different optimization levels (the classic one is instruction
+  reordering causing concurrent access bugs!)
+
+Because your debug (testing) build is not running the same code, you may see
+these bugs only show up during field test, rather than when the device is on
+your desk and connected to a debugger!
+
+It's also generally not desirable to adjust compiler optimization level when a
+project is nearing time to ship, to decrease risk. If the debug + release builds
+are using the same optimization settings, it's less likely to hit a surprise
+out-of-space issue that can stop a project in its tracks 😱.
+
+**Summary:**
+
+- use the same optimization settings for all builds to decrease risk
+
+## `NDEBUG` preprocessor definition
+
+Often these Debug/Release builds will have some preprocessor defined symbols
+such as:
+
+- `DEBUG=1` (debug enabled)
+- `NDEBUG=0` (indicating this is NOT a non-debug build)
+
+Usually these flags will be used to adjust some of the following features for
+production builds:
+
+1. Changing **log verbosity** (disabling DEBUG or INFO levels in production, to
+   prevent unnecessary flash wear)
+   - I'd recommend instead, sticking with the **production** level, and in
+     exceptional cases (problematic device), upgrade to a more verbose level
+     temporarily
+   - Depending on the particular system, it may be preferable to adjust
+     verbosity at runtime, and keep the build the same. For example, leave
+     WARNING and ERROR enabled by default, but allow increasing verbosity via a
+     persistent on-device setting, or a temporary runtime flag.
+   - Be especially wary of builds with different log verbosity- it can impact
+     timing and memory usage _dramatically_!
+2. Disabling an **"engineering interface"** (eg ssh or local serial console
+   backdoor)
+   - This can be a security requirement. An alternative is to build
+     authentication into the engineering console, but this can be error prone so
+     proceed cautiously.
+   - Having the interface available (in an authenticated way) on production
+     units can be extremely useful when diagnosing faulty customer units that
+     have been returned for analysis!
+3. Enabling **signed/encrypted builds**
+   - Also very tricky- you definitely want to be verifying the
+     signing/encryption implementation works throughout the development and test
+     cycle, instead of causing surprises at the end (execute-in-place encryption
+     too slow, key management not actually production ready, etc).
```

Comment:
```suggestion
   - Also very tricky! Best practice is to verify the
     signing/encryption implementation works throughout the development and test
     cycle, instead of causing surprises at the end (execute-in-place encryption
     too slow, key management not actually production ready, etc).
```